README.md
# LaTeX Beamer Material Theme

<p align="center">
  <a href="demo/lightblue.pdf">
    <img src="demo/lightblue-01.png?raw=true">
  </a>
  <a href="demo/lightblue.pdf">
    <img src="demo/lightblue-02.png?raw=true">
  </a>
  <a href="demo/darkpurple.pdf">
    <img src="demo/darkpurple-02.png?raw=true">
  </a>
  <a href="demo/lightblue.pdf">
    <img src="https://github.com/peterpf/beamer-material-theme/blob/develop/demo/lightblue-03.png?raw=true">
  </a>
</p>

## Setup

`LuaTex` and the `roboto` font are required to compile the presentation.

As an alternative, use the included `docker` container to produce a pdf file.
Execute following commands in the directory of the cloned repository:

```bash
docker build . -t latex-beamer
docker run -v $PWD:/data/ latex-beamer example.tex
```

Replace `example.tex` with the main tex file.

## Include as git submodule

Go to your working directory and clone the `beamer-material-theme` repository into the folder `theme` with

```bash
git submodule add git@github.com:peterpf/beamer-material-theme.git theme
```

Then do following:

- Open [src/beamerthemematerial.sty](src/beamerthemematerial.sty) and adapt the paths for the styles:

  ```latex
  \usepackage{theme/src/material_colors}
  \usepackage{theme/src/beamerinnerthemematerial}
  \usepackage{theme/src/beamerouterthemematerial}
  \usepackage{theme/src/beamercolorthemematerial}
  ```

- Copy [example.tex](example.tex) from `src` folder to `theme` folder:

    ```bash
    .
    ├── theme
    |   ├── src/
    |   ├── example.tex
    ├── example.tex
    ├── ...
    ```

- Open the copied file and look for the line that includes the theme:

  ```latex
  \usepackage{src/beamerthemematerial} % replace this
  \usepackage{src/beamerthemematerial} % with this
  ```

- Use the copied file as a starting point for your presentation.
- Execute the docker command mentioned above in the project root folder.

## Theme Configuration

Colors can be adapted via the [config.lua](src/config.lua).
Undefined complementary colors will be calculated based on their respective main colors.
Predefined themes are already there, choose the one you want by changing the return value of [config.lua](src/config.lua), or create your own:

```lua
return lightBlue
-- or
return darkPurple
```

## Important Notes

- Packages that should not be included
  - `graphicx` or `graphics` as they are already included in beamer
  - `transparency`, `svg` as it messes with the tikz package's `fill opacity` (shadows won't be displayed correctly)

## Shoutouts

- [docker-luatex](https://github.com/brokenpylons/docker-lualatex)
- [docker-texlive](https://github.com/thomasWeise/docker-texlive)
- [Lua Colors Library](https://github.com/yuri/lua-colors)


CHANGELOG.md
# Changelog

## 3.0.0

### Added

- Dockerization of the build-process
- `darkpurple` theme to `config.lua`

### Changed

- build-system `rake` to `LuaTeX` (removed python code and replaced it with lua)
- `config.json` to `config.lua`

## 2.1.0

### Added

- Command `\topicFramePrimary[subtitle]{Title}` which creates a topic-introduction frame with primary color as background
- Command `\topicFrameSecondary[subtitle]{Title}` which creates a topic-introduction frame with secondary color as background

## 2.0.1

### Added

- Page number to navigation symbols
- Example pdf

### Changed

- Caption color
- `\sidebysideX` layout's first optional argument specifies the ratio of the sides
- Updated readme + example slides

## 2.0.0

### Added

- Side by side slides
- Build system

### Fixed

- Shadow was not displayed correctly in some cases, added a guideline

### Changed

- Folder structure improvements
- Theme generation is now done with

    ```bash
    rake setup
    rake default
    ```

- Theme settings are done in the `config.json` file

## 1.0.0

Initial release featuring:

- Theme customization via `LaTeX` commands
- Material Design


src/Utils.lua
local ColorUtils = require("lib/colors")
local Utils = {}

-- Convert hex color representation to {r, g, b}, all values are in range 0-1.
-- Source: Somewhere on stackoverflow
function hex2rgb (hex)
  local hex = hex:gsub("#","")
  if hex:len() == 3 then
    return (tonumber("0x"..hex:sub(1,1))*17)/255, (tonumber("0x"..hex:sub(2,2))*17)/255, (tonumber("0x"..hex:sub(3,3))*17)/255
  else
    return tonumber("0x"..hex:sub(1,2))/255, tonumber("0x"..hex:sub(3,4))/255, tonumber("0x"..hex:sub(5,6))/255
  end
end

-- Convert {r, g, b} to a hex string. All values must be in range 0 - 255.
-- Source: https://gist.github.com/marceloCodget/3862929
function rgb2hex(rgb)
  local hexadecimal = ""
	for key, value in pairs(rgb) do
		local hex = ""

		while(value > 0)do
      local index = math.floor(value%16 + 1)
      value = math.floor(value / 16)
			hex = string.sub("0123456789ABCDEF", index, index) .. hex
		end

		if(string.len(hex) == 0)then
			hex = "00"

		elseif(string.len(hex) == 1)then
			hex = "0" .. hex
		end

		hexadecimal = hexadecimal .. hex
	end

	return hexadecimal
end

-- Convert a hex string to a Color object
function hex2color(hex)
  local r,g,b = hex2rgb(hex)
  local h,s,l = ColorUtils.rgb_to_hsl(r,g,b)
  return ColorUtils.new(h,s,l)
end

-- Represent a Color object in hex
function color2hex(color)
  local nr, ng, nb = ColorUtils.hsl_to_rgb(color.H, color.S, color.L)
  return rgb2hex({nr*255, ng*255, nb*255})
end

-- Multiply Luminance value of given color with factor
function Utils.changeColorBrightness(hexColor, factor)
  local color = hex2color(hexColor)
  color = color:lighten_by(factor)
  return color2hex(color)
end

-- Return the color (brighterHexColor, darkerHexColor) which is a better contrast fit for the given hexColor
function Utils.getContrastingColor(hexColor, brighterHexColor, darkerHexColor)
  local color = hex2color(hexColor)
  if color.L < 0.6 then
    return brighterHexColor
  else
    return darkerHexColor
  end
end

-- Return the corresponding LaTeX code to define a new HTML color
function Utils.defineHTMLColor(colorName, hexColor)
  tex.sprint("\\definecolor{",colorName, "}{HTML}{",hexColor, "}")
end

-- Perform checks on the colors and define missing optional colors.
function Utils.checkColors(config)
  -- Factors by which colors get brightened/darkened
  local FACTOR_BRIGHTEN = 1.8
  local FACTOR_DARKEN = 0.4

  -- Fallback colors if TextColors are not defined
  local BrightTextColor = Utils.changeColorBrightness(config.PrimaryColor, FACTOR_BRIGHTEN)
  local DarkTextColor = Utils.changeColorBrightness(config.PrimaryColor, FACTOR_BRIGHTEN)

  -- Define optional colors if they are missing
  if config.PrimaryColorLight == nil then
    config.PrimaryColorLight = Utils.changeColorBrightness(config.PrimaryColor, FACTOR_DARKEN)
  end
  if config.PrimaryColorDark == nil then
    config.PrimaryColorDark = Utils.changeColorBrightness(config.PrimaryColor, FACTOR_DARKEN)
  end
  if config.PrimaryTextColor == nil then
    config.PrimaryTextColor = Utils.getContrastingColor(config.PrimaryColor, BrightTextColor, DarkTextColor)
  end
  if config.SecondaryColorLight == nil then
    config.SecondaryColorLight = Utils.changeColorBrightness(config.SecondaryColor, FACTOR_BRIGHTEN)
  end
  if config.SecondaryColorDark == nil then
    config.SecondaryColorDark = Utils.changeColorBrightness(config.SecondaryColor, FACTOR_DARKEN)
  end
  if config.SecondaryTextColor == nil then
    config.SecondaryTextColor = Utils.getContrastingColor(config.SecondaryColor, BrightTextColor, DarkTextColor)
  end
  return config
end

return Utils

lib/colors.lua
-- Original Source: https://github.com/yuri/lua-colors

-----------------------------------------------------------------------------
-- Provides support for color manipulation in HSL color space.
--
-- http://sputnik.freewisdom.org/lib/colors/
--
-- License: MIT/X
--
-- (c) 2008 Yuri Takhteyev (yuri@freewisdom.org) *
--
-- * rgb_to_hsl() implementation was contributed by Markus Fleck-Graffe.
-----------------------------------------------------------------------------

local M = {}

local Color = {}
local Color_mt = {__metatable = {}, __index = Color}

local rgb_string_to_hsl -- defined below

-----------------------------------------------------------------------------
-- Instantiates a new "color".
--
-- @param H              hue (0-360) _or_ an RGB string ("#930219")
-- @param S              saturation (0.0-1.0)
-- @param L              lightness (0.0-1.0)
-- @return               an instance of Color
-----------------------------------------------------------------------------
local function new(H, S, L)
   if type(H) == "string" and H:sub(1,1)=="#" and H:len() == 7 then
      H, S, L = rgb_string_to_hsl(H)
   end
   assert(Color_mt)
   return setmetatable({H = H, S = S, L = L}, Color_mt)
end
M.new = new

-----------------------------------------------------------------------------
-- Converts an HSL triplet to RGB
-- (see http://homepages.cwi.nl/~steven/css/hsl.html).
--
-- @param H              hue (0-360)
-- @param S              saturation (0.0-1.0)
-- @param L              lightness (0.0-1.0)
-- @return               an R, G, and B component of RGB
-----------------------------------------------------------------------------

local function hsl_to_rgb(h, s, L)
   h = h/360
   local m1, m2
   if L<=0.5 then
      m2 = L*(s+1)
   else
      m2 = L+s-L*s
   end
   m1 = L*2-m2

   local function _h2rgb(m1, m2, h)
      if h<0 then h = h+1 end
      if h>1 then h = h-1 end
      if h*6<1 then
         return m1+(m2-m1)*h*6
      elseif h*2<1 then
         return m2
      elseif h*3<2 then
         return m1+(m2-m1)*(2/3-h)*6
      else
         return m1
      end
   end

   return _h2rgb(m1, m2, h+1/3), _h2rgb(m1, m2, h), _h2rgb(m1, m2, h-1/3)
end
M.hsl_to_rgb = hsl_to_rgb

-----------------------------------------------------------------------------
-- Converts an RGB triplet to HSL.
-- (see http://easyrgb.com)
--
-- @param r              red (0.0-1.0)
-- @param g              green (0.0-1.0)
-- @param b              blue (0.0-1.0)
-- @return               corresponding H, S and L components
-----------------------------------------------------------------------------

local function rgb_to_hsl(r, g, b)
   --r, g, b = r/255, g/255, b/255
   local min = math.min(r, g, b)
   local max = math.max(r, g, b)
   local delta = max - min

   local h, s, l = 0, 0, ((min+max)/2)

   if l > 0 and l < 0.5 then s = delta/(max+min) end
   if l >= 0.5 and l < 1 then s = delta/(2-max-min) end

   if delta > 0 then
      if max == r and max ~= g then h = h + (g-b)/delta end
      if max == g and max ~= b then h = h + 2 + (b-r)/delta end
      if max == b and max ~= r then h = h + 4 + (r-g)/delta end
      h = h / 6;
   end

   if h < 0 then h = h + 1 end
   if h > 1 then h = h - 1 end

   return h * 360, s, l
end
M.rgb_to_hsl = rgb_to_hsl

-- already local, see at the bottom
function rgb_string_to_hsl(rgb)
   return rgb_to_hsl(tonumber(rgb:sub(2,3), 16)/255,
                     tonumber(rgb:sub(4,5), 16)/255,
                     tonumber(rgb:sub(6,7), 16)/255)
end
M.rgb_string_to_hsl = rgb_string_to_hsl

-----------------------------------------------------------------------------
-- Converts the color to an RGB string.
--
-- @return               a 6-digit RGB representation of the color prefixed
--                       with "#" (suitable for inclusion in HTML)
-----------------------------------------------------------------------------

function Color:to_rgb()
   local r, g, b = hsl_to_rgb(self.H, self.S, self.L)
   local rgb = {hsl_to_rgb(self.H, self.S, self.L)}
   local buffer = "#"
   for i,v in ipairs(rgb) do
	  buffer = buffer..string.format("%02x",math.floor(v*255+0.5))
   end
   return buffer
end

-----------------------------------------------------------------------------
-- Creates a new color with hue different by delta.
--
-- @param delta          a delta for hue.
-- @return               a new instance of Color.
-----------------------------------------------------------------------------
function Color:hue_offset(delta)
   return new((self.H + delta) % 360, self.S, self.L)
end

-----------------------------------------------------------------------------
-- Creates a complementary color.
--
-- @return               a new instance of Color
-----------------------------------------------------------------------------
function Color:complementary()
   return self:hue_offset(180)
end

-----------------------------------------------------------------------------
-- Creates two neighboring colors (by hue), offset by "angle".
--
-- @param angle          the difference in hue between this color and the
--                       neighbors
-- @return               two new instances of Color
-----------------------------------------------------------------------------
function Color:neighbors(angle)
   local angle = angle or 30
   return self:hue_offset(angle), self:hue_offset(360-angle)
end

-----------------------------------------------------------------------------
-- Creates two new colors to make a triadic color scheme.
--
-- @return               two new instances of Color
-----------------------------------------------------------------------------
function Color:triadic()
   return self:neighbors(120)
end

-----------------------------------------------------------------------------
-- Creates two new colors, offset by angle from this colors complementary.
--
-- @param angle          the difference in hue between the complementary and
--                       the returned colors
-- @return               two new instances of Color
-----------------------------------------------------------------------------
function Color:split_complementary(angle)
   return self:neighbors(180-(angle or 30))
end

-----------------------------------------------------------------------------
-- Creates a new color with saturation set to a new value.
--
-- @param saturation     the new saturation value (0.0 - 1.0)
-- @return               a new instance of Color
-----------------------------------------------------------------------------
function Color:desaturate_to(saturation)
   return new(self.H, saturation, self.L)
end

-----------------------------------------------------------------------------
-- Creates a new color with saturation set to a old saturation times r.
--
-- @param r              the multiplier for the new saturation
-- @return               a new instance of Color
-----------------------------------------------------------------------------
function Color:desaturate_by(r)
   return new(self.H, math.min(self.S*r, 1), self.L)
end

-----------------------------------------------------------------------------
-- Creates a new color with lightness set to a new value.
--
-- @param lightness      the new lightness value (0.0 - 1.0)
-- @return               a new instance of Color
-----------------------------------------------------------------------------
function Color:lighten_to(lightness)
   return new(self.H, self.S, lightness)
end

-----------------------------------------------------------------------------
-- Creates a new color with lightness set to a old lightness times r.
--
-- @param r              the multiplier for the new lightness
-- @return               a new instance of Color
-----------------------------------------------------------------------------
function Color:lighten_by(r)
   return new(self.H, self.S, math.min(self.L*r, 1))
end

-----------------------------------------------------------------------------
-- Creates n variations of this color using supplied function and returns
-- them as a table.
--
-- @param f              the function to create variations
-- @param n              the number of variations
-- @return               a table with n values containing the new colors
-----------------------------------------------------------------------------
function Color:variations(f, n)
   n = n or 5
   local results = {}
   for i=1,n do
	  table.insert(results, f(self, i, n))
   end
   return results
end

-----------------------------------------------------------------------------
-- Creates n tints of this color and returns them as a table
--
-- @param n              the number of tints
-- @return               a table with n values containing the new colors
-----------------------------------------------------------------------------
function Color:tints(n)
   local f = function (color, i, n) 
                return color:lighten_to(color.L + (1-color.L)/n*i)
             end
   return self:variations(f, n)
end

-----------------------------------------------------------------------------
-- Creates n shades of this color and returns them as a table
--
-- @param n              the number of shades
-- @return               a table with n values containing the new colors
-----------------------------------------------------------------------------
function Color:shades(n)
   local f = function (color, i, n) 
                return color:lighten_to(color.L - (color.L)/n*i)
             end
   return self:variations(f, n)
end

function Color:tint(r)
      return self:lighten_to(self.L + (1-self.L)*r)
end

function Color:shade(r)
      return self:lighten_to(self.L - self.L*r)
end

Color_mt.__tostring = Color.to_rgb

-- allow to use `colors(...)` instead of `colors.new(...)`
setmetatable(M, {__call=function(_, ...) return new(...) end})

return M


src/config.lua
local lightBlue =  {
    -- Required
    PrimaryColor            = "2196f3";
    SecondaryColor          = "fafafa";
    AccentColor             = "80d8ff";
    Shadow                  = "000000";
    BackgroundColor1        = "e0e0e0";
    BackgroundColor2        = "f5f5f5";
    BackgroundColor3        = "fafafa";
    BackgroundColor4        = "ffffff";

    -- Optional
    PrimaryColorLight       = "87fbff";
    PrimaryColorDark        = "002f8c";
    PrimaryTextColor        = "edffff";
    SecondaryColorLight     = "ffffff";
    SecondaryColorDark      = "949494";
    SecondaryTextColor      = "00000d";
}
local darkPurple =  {
    -- Required
    PrimaryColor            = "7b1fa2";
    SecondaryColor          = "fb8c00";
    AccentColor             = "d500f9";
    Shadow                  = "000000";
    BackgroundColor1        = "e0e0e0";
    BackgroundColor2        = "f5f5f5";
    BackgroundColor3        = "fafafa";
    BackgroundColor4        = "ffffff";

    -- Optional
    PrimaryColorLight       = "ae52d4";
    PrimaryColorDark        = "4a0072";
    PrimaryTextColor        = "ffffff";
    SecondaryColorLight     = "ffbd45";
    SecondaryColorDark      = "c25e00";
    SecondaryTextColor      = "000000";
}

-- Choose one of the above color themes
return lightBlue

src/settings.lua
local Utils = require("src/Utils")
local config = require("src/config")

local baseConfig = Utils.checkColors(config)

print(baseConfig)

return {
    PrimaryColor            = baseConfig.PrimaryColor;
    PrimaryColorLight       = baseConfig.PrimaryColorLight;
    PrimaryColorDark        = baseConfig.PrimaryColorDark;
    PrimaryTextColor        = baseConfig.PrimaryTextColor;
    SecondaryColor          = baseConfig.SecondaryColor;
    SecondaryColorLight     = baseConfig.SecondaryColorLight;
    SecondaryColorDark      = baseConfig.SecondaryColorDark;
    SecondaryTextColor      = baseConfig.SecondaryTextColor;
    AccentColor             = baseConfig.AccentColor;
    Shadow                  = baseConfig.Shadow;
    BackgroundColor1        = baseConfig.BackgroundColor1;
    BackgroundColor2        = baseConfig.BackgroundColor1;
    BackgroundColor3        = baseConfig.BackgroundColor3;
    BackgroundColor4        = baseConfig.BackgroundColor3;
}

build_theme.py
#!/usr/bin/env python3
import json
import os
from os import listdir
from os.path import isfile, join
from colour import Color

__location__ = os.path.realpath(
    os.path.join(os.getcwd(), os.path.dirname(__file__)))
config_filepath = os.path.join(__location__, 'config.json')
latex_source_filepath = os.path.join(__location__, 'dist')
latex_source_files = [join(latex_source_filepath, f) for f in listdir(latex_source_filepath) if isfile(join(latex_source_filepath, f))]


with open(config_filepath, 'r') as configFile:
    configJson = configFile.read()

config = json.loads(configJson)

config_map = {
    'primaryColor': {
        'color': None,
        'replace_string': '%PRIMARY_COLOR%'
    },
    'primaryColorLight': {
        'color': None,
        'replace_string': '%PRIMARY_COLOR_LIGHT%'
    },
    'primaryColorDark': {
        'color': None,
        'replace_string': '%PRIMARY_COLOR_DARK%'
    },
    'primaryTextColor': {
        'color': None,
        'replace_string': '%PRIMARY_TEXT_COLOR%'
    },
    'secondaryColor': {
        'color': None,
        'replace_string': '%SECONDARY_COLOR%'
    },
    'secondaryColorLight': {
        'color': None,
        'replace_string': '%SECONDARY_COLOR_LIGHT%'
    },
    'secondaryColorDark': {
        'color': None,
        'replace_string': '%SECONDARY_COLOR_DARK%'
    },
    'secondaryTextColor': {
        'color': None,
        'replace_string': '%SECONDARY_TEXT_COLOR%'
    },
    'accentColor': {
        'color': None,
        'replace_string': '%ACCENT_COLOR%'
    },
    'backgroundColorShadow': {
        'color': None,
        'replace_string': '%BACKGROUND_COLOR_SHADOW%'
    },
    'backgroundColor1': {
        'color': None,
        'replace_string': '%BACKGROUND_COLOR_1%'
    },
    'backgroundColor2': {
        'color': None,
        'replace_string': '%BACKGROUND_COLOR_2%'
    },
    'backgroundColor3': {
        'color': None,
        'replace_string': '%BACKGROUND_COLOR_3%'
    },
    'backgroundColor4': {
        'color': None,
        'replace_string': '%BACKGROUND_COLOR_4%'
    },
    
}


def hex_to_rgb(value):
    value = value.lstrip('#')
    return list(int(value[i:i+2], 16) for i in (0, 2, 4))

def rgb_to_hex(r, g, b):
    return '#%02x%02x%02x' % (r, g, b)

def rgb_brighten(r, g, b, percentage):
    return (max(r + percentage, 255), max(g + percentage, 255), max(b + percentage, 255))

def rgb_darken(r, g, b, percentage):
    return (max(r - percentage, 0), max(g - percentage, 0), max(b - percentage, 0))


''' Complement the colours in a RGB image 

    Written by PM 2Ring 2016.10.08
'''
# Sum of the min & max of (a, b, c)
def hilo(a, b, c):
    if c < b: b, c = c, b
    if b < a: a, b = b, a
    if c < b: b, c = c, b
    return a + c

def complement(r, g, b):
    k = hilo(r, g, b)
    return tuple(k - u for u in (r, g, b))

def color_brighten(color, percentage):
    r = int(255 * color.red)
    g = int(255 * color.green)
    b = int(255 * color.blue)
    p = int(255 * percentage)
    return Color(rgb=(
        min(r + p, 255) / 255,
        min(g + p, 255) / 255,
        min(b + p, 255) / 255)
        )

def color_darken(color, percentage):
    r = int(255 * color.red)
    g = int(255 * color.green)
    b = int(255 * color.blue)
    p = int(255 * percentage)
    return Color(rgb=(
        max(r - p, 0) / 255,
        max(g - p, 0) / 255,
        max(b - p, 0) / 255)
        )


def replace_in_files(replace_text, replace_value):
    for source_file in latex_source_files:
        file_content = open(source_file, 'r').read()
        # print('reading file: ' + source_file)
        if replace_text in file_content:
            #print('replacing ' + replace_text + ' w/ ' + replace_value)
            open(source_file, 'w').write(file_content.replace(replace_text, replace_value))

# Check if first color is darker than second color
saturation_equivalence = lambda c1, c2: c1.saturation < c2.saturation

def read_color_from_config(color_name):
    hex_value = config[color_name]
    if hex_value is None:
        return None
    return Color(str(hex_value), equality=saturation_equivalence)

def get_config_map_color(color_name):
    return config_map[color_name]['color']

def set_config_map_color(color_name, color_value):
    config_map[color_name]['color'] = color_value

def get_contrast_color(color, bright_color, dark_color):
    return bright_color if color.get_luminance() < 0.6 else dark_color


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ~ Read color values from json file ~
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

for colorName, colorObject in config_map.items():
    colorObject['color'] = read_color_from_config(colorName)

primaryColor = get_config_map_color('primaryColor')
primaryColorLight = get_config_map_color('primaryColorLight')
primaryColorDark = get_config_map_color('primaryColorDark')
primaryTextColor = get_config_map_color('primaryTextColor')
secondaryColor = get_config_map_color('secondaryColor')
secondaryColorLight = get_config_map_color('secondaryColorLight')
secondaryColorDark = get_config_map_color('secondaryColorDark')
secondaryTextColor = get_config_map_color('secondaryTextColor')

if primaryColor is None:
    raise('primaryColor is not defined')
if secondaryColor is None:
    raise('secondaryColor is not defined')

brightTextColor = color_brighten(primaryColor, 0.8)
darkTextColor = color_darken(primaryColor, 0.9)

if primaryColorLight is None:
    set_config_map_color('primaryColorLight', color_brighten(primaryColor, 0.4))
if primaryColorDark is None:
    set_config_map_color('primaryColorDark', color_darken(primaryColor, 0.4))
if primaryTextColor is None:
    set_config_map_color('primaryTextColor', get_contrast_color(primaryColor, brightTextColor, darkTextColor))

if secondaryColorLight is None:
    set_config_map_color('secondaryColorLight', color_brighten(secondaryColor, 0.4))
if secondaryColorDark is None:
    set_config_map_color('secondaryColorDark', color_darken(secondaryColor, 0.4))
if secondaryTextColor is None:
    set_config_map_color('secondaryTextColor', get_contrast_color(secondaryColor, brightTextColor, darkTextColor))


for colorName, colorObject in config_map.items():
    color = get_config_map_color(colorName)
    replace_string = colorObject['replace_string']
    if color is None:
        print(f"The color '{colorName}' is still undefined")
        raise("Exception")
    # color_rgb = hex_to_rgb(color.hex.lstrip('#'))
    # color_complement_rgb = complement(*color_rgb)
    # color_complement_hex = rgb_to_hex(*color_complement_rgb)
    # print(json_attribute_name + ' '+ color_complement_hex)
    replace_in_files(replace_string, color.hex_l.lstrip('#'))